torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1197,6 +1200,32 @@ void ProcessGroupNCCL::heartbeatMonitor() {
     // dump debugging info, and we avoid hammering the TCPStore from all PGs on
     // the same rank.
     if (checkTimeoutSignal) {
+      // There are two scenarios where monitor thread will dump on timeout:
+      // 1. The local rank is the first to observe a timeout.shouldDump_ will be
+      // set to true.
+      // 2. other ranks detected the timeout and signal the local rank to dump
+      // In addtion, monitor threads will dump if watchdog threads has no
+      // heartbeat.
+      if (shouldDump_.load()) {
+        errorMsg = c10::str(
+            logPrefix(),
+            "Received a timeout signal from this local rank and will ",
+            "start to dump the debug info. ",
+            "Last enqueued NCCL work: ",
+            lastEnqueuedSeq_,
+            ", last completed NCCL work: ",
+            lastCompletedSeq_,
+            ".");
+        exitMsg = c10::str(
+            "ProcessGroupNCCL's watchdog detected a collective timeout from the local rank. ",
+            "This is most likely caused by incorrect usages of collectives, e.g., wrong ",
+            "sizes used across ranks, the order of collectives is not same for all ranks ",
+            "or the scheduled collective, for some reason, didn't run. Additionally, ",
+            "this can be caused by GIL deadlock or other reasons such as network errors or ",
+            "bugs in the communications library (e.g. NCCL), etc. We tried our best to ",
+            "dump the debug info into the storage to help you debug the issue.");
+        break;
```

Comment:
does this break mean 'exit the monitor loop and kill the job?'

bc that isn't safe.  When the 'dumpPipe' is invoked, we expect the dump signal to propagate here and dump, but the training should continue.  Only if the dump is caused by timeout we want to begin the teardown.

Reply:
Oh, didn't know dumpPipe is for dump and continue

Reply:
yea. its probably worth adding a test that after dumpPipe is written the process doesn't die. i thought we had that actually.

Reply:
We have the test, but I misunderstood the dump logic, so to disable the Kill part through: os.environ['TORCH_NCCL_ENABLE_MONITORING'] = '0', in the next version, I will remove this env change